### DISCUSSION TITLE
Can I build a faster than Cholesky random generator with Gpytorch? (not working at the moment)

### DISCUSSION THREAD

**Discussion by nrigol** [cornellius-gp/gpytorch#2471] [Q&A]
I am actually generating correlated variables using numpy cholesky, I would like to use gpytorch to speed up the process, the maximum size of the matrix is not big (200x200) but I am calculating the correlated variable in a loop so the operation is repeated at every temporal iteration in my code.

`
def generate_correlated_variables(mean, covariance_matrix):
    cholesky_matrix = np.linalg.cholesky(covariance_matrix)
    normal_random_variables = np.random.normal(size=[mean.shape[0],mean.shape[1]])
    correlated_variables = mean + np.dot(normal_random_variables, cholesky_matrix.T)
    return correlated_variables
`
    
my idea was to use :
`
eta_dist = MultivariateNormal(torch.zeros_like(mean), covariance_matrix)
samples = eta_dist.sample()
`

but it is actually much slower than using numpy. 
Do you have any explanation for this ?
(My understanding is that for a matrix of my current size I am still using Cholesky but torch.Cholesky, am I right?)
Do you have any recommendation on alternatives I could use to speed up the process ?

**Comment by Balandat** [comment]
Yes this should still use `torch.lialg.cholesky` under the hood. How big is the time difference? How are you timing this? I'm not surprised that the numpy version is faster (at least on the CPU this will use the same underlying numerical linear algebra libraries to compute the matrix decomposition) since torch will have more overhead.
